File: 448.find-all-numbers-disappeared-in-an-array.java
/*
 * @lc app=leetcode id=448 lang=java
 *
 * [448] Find All Numbers Disappeared in an Array
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        List<Integer> ret = new ArrayList<>();
        for (int ix = 0; ix != nums.length; ++ix) {
            int idx = Math.abs(nums[ix]) - 1;
            if (nums[idx] > 0) {
                nums[idx] = -nums[idx];
            }
        }
        for (int ix = 0; ix != nums.length; ++ix) {
            if (nums[ix] > 0) ret.add(ix + 1);
        }
        return ret;
    }
}
// @lc code=end

File: 448.find-all-numbers-disappeared-in-an-array.cpp
/*
 * @lc app=leetcode id=448 lang=cpp
 *
 * [448] Find All Numbers Disappeared in an Array
 */

// @lc code=start
#include <cstdlib>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        vector<int> ret;
        for (auto ix = 0ull; ix != nums.size(); ++ix) {
            auto key = abs(nums[ix]) - 1;
            if (nums[key] > 0) nums[key] = -nums[key];
        }
        for (auto ix = 0ull; ix != nums.size(); ++ix) {
            if (nums[ix] > 0) ret.push_back(ix + 1);
        }
        return ret;
    }
};
// @lc code=end
